pkg/cloud/vsphere/utils/utils.go

```diff
@@ -1,31 +1,304 @@
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/constants"
+	"sigs.k8s.io/cluster-api-provider-vsphere/pkg/cloud/vsphere/services/certificates"
 )
 
+const (
+	defaultBindPort = 6443
+)
+
+// GetMachineRole returns either "controlplane", "node", or an empty
+// string if the role could be ascertained.
+func GetMachineRole(machine *clusterv1.Machine) string {
+	set := machine.Labels["set"]
+	if ok, _ := regexp.MatchString(`(?i)controlplane|master`, set); ok {
+		return "controlplane"
+	}
+	if ok, _ := regexp.MatchString(`(?i)(?:worker-)?node`, set); ok {
+		return "node"
+	}
+	nodeType := machine.Labels["node-type"]
+	if ok, _ := regexp.MatchString(`(?i)(?:worker-)?node`, nodeType); ok {
+		return "node"
+	}
+	return ""
+}
+
+// byMachineCreatedTimestamp implements sort.Interface for []clusterv1.Machine
+// based on the machine's creation timestamp.
+type byMachineCreatedTimestamp []*clusterv1.Machine
+
+func (a byMachineCreatedTimestamp) Len() int      { return len(a) }
+func (a byMachineCreatedTimestamp) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a byMachineCreatedTimestamp) Less(i, j int) bool {
+	return a[i].CreationTimestamp.Before(&a[j].CreationTimestamp)
+}
+
+// GetControlPlaneEndpoint returns the control plane endpoint for the cluster.
+// If a control plane endpoint was specified in the cluster configuration, then
+// that value will be returned.
+// Otherwise this function will return the endpoint of the first control plane
+// node in the cluster that reports an IP address.
+// If no control plane nodes have reported an IP address then this function
+// returns an error.
+func GetControlPlaneEndpoint(
+	cluster *clusterv1.Cluster,
+	client clientv1.Interface) (string, error) {
+
+	if len(cluster.Status.APIEndpoints) > 0 {
+		controlPlaneEndpoint := net.JoinHostPort(
+			cluster.Status.APIEndpoints[0].Host,
+			strconv.Itoa(cluster.Status.APIEndpoints[0].Port))
+		klog.V(2).Infof("got control plane endpoint from cluster APIEndpoints "+
+			" %s=%s %s=%s %s=%s",
+			"control-plane-endpoint", controlPlaneEndpoint,
+			"cluster-namespace", cluster.Namespace,
+			"cluster-name", cluster.Name)
+		return controlPlaneEndpoint, nil
+	}
+
+	clusterProviderConfig, err := GetClusterProviderSpec(cluster.Spec.ProviderSpec)
+	if err != nil {
+		return "", errors.Wrapf(
+			err,
+			"unable to get cluster provider config while searching for "+
+				"control plane endpoint %s=%s %s=%s",
+			"cluster-name", cluster.Name,
+			"cluster-namespace", cluster.Namespace)
+	}
+	if controlPlaneEndpoint := clusterProviderConfig.ClusterConfiguration.ControlPlaneEndpoint; controlPlaneEndpoint != "" {
+		klog.V(2).Infof("got control plane endpoint from cluster config %s=%s %s=%s %s=%s",
+			"control-plane-endpoint", controlPlaneEndpoint,
+			"cluster-namespace", cluster.Namespace,
+			"cluster-name", cluster.Name)
+		return controlPlaneEndpoint, nil
+	}
+
+	if client == nil {
```

Comment:
nit: move this to the top of the function?

Reply:
The first three functions don't require `client` and can return the result without it. That's why it's at this location. Here and beyond `client` is required, but not before. And there are callers that send in `nil` clients by design.

Reply:
But if we return error anyways when client is nil isn't it redundant to even call the first 3 functions?

Reply:
The first three functions can return the ControlPlaneEndpoint without the client. Beyond that point the client is needed.